Sources/SourceKitLSP/Swift/CodeActions/ConvertJSONToCodableStruct.swift

```diff
@@ -0,0 +1,318 @@
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import LanguageServerProtocol
+import SwiftRefactor
+import SwiftSyntax
+
+/// Convert JSON literals into corresponding Swift structs that conform to the
+/// `Codable` protocol.
+///
+/// ## Before
+///
+/// ```javascript
+/// {
+///     "name": "Produce",
+///     "shelves": [
+///         {
+///             "name": "Discount Produce",
+///             "product": {
+///                 "name": "Banana",
+///                 "points": 200,
+///                 "description": "A banana that's perfectly ripe."
+///             }
+///         }
+///     ]
+/// }
+/// ```
+///
+/// ## After
+///
+/// ```swift
+/// struct JSONValue: Codable {
+///   var name: String
+///   var shelves: [Shelves]
+///
+///   struct Shelves: Codable {
+///     var name: String
+///     var product: Product
+///
+///     struct Product: Codable {
+///       var description: String
+///       var name: String
+///       var points: Double
+///     }
+///   }
+/// }
+/// ```
+@_spi(Testing)
+public struct ConvertJSONToCodableStruct: EditRefactoringProvider {
+  @_spi(Testing)
+  public static func textRefactor(
+    syntax: Syntax,
+    in context: Void
+  ) -> [SourceEdit] {
+    // Dig out a syntax node that looks like it might be JSON or have JSON
+    // in it.
+    guard let preflight = preflightRefactoring(syntax) else {
+      return []
+    }
+
+    // Dig out the text that we think might be JSON.
+    let text: String
+    switch preflight {
+    case let .closure(closure):
+      text = closure.trimmedDescription
+    case let .tail(closure, unexpected):
+      text = closure.trimmedDescription + unexpected.description
+    case .stringLiteral(_, let literalText):
+      text = literalText
+    }
+
+    // Try to process this as JSON.
+    guard
+      let object = try? JSONSerialization.jsonObject(with: text.data(using: .utf8)!),
+      let serial = object as? Dictionary<String, Any>
+    else {
+      return []
+    }
+
+    // Build the declarations from the JSON dictionary.
+    let decls = build(from: serial)
+
+    // Render the change into
+    switch preflight {
+    case .closure(let closure):
+      return [
+        SourceEdit(range: closure.trimmedRange, replacement: decls.description)
+      ]
+    case .tail(let closure, let unexpected):
+      return [
+        SourceEdit(range: closure.trimmedRange, replacement: decls.description),
+        SourceEdit(range: unexpected.range, replacement: ""),
+      ]
+    case .stringLiteral(let literal, _):
+      return [
+        SourceEdit(
+          range: literal.endPosition..<literal.endPosition,
+          replacement: "\n" + decls.description
+        )
+      ]
+    }
+  }
+}
+
+extension ConvertJSONToCodableStruct {
```

Comment:
Let’s just shove them all together instead of having `extension`s of the same type within the same file. 

```suggestion

```

Same once more a few lines below.

Reply:
I want the protocol-conforming extensions separate, but other than that---sure

Reply:
This is no longer a protocol-conforming extension 😜 Also, I’m not generally a fan of putting protocol conformances in extensions. I think it makes it harder to reason about which protocols a type conforms to.